### PR TITLE
refactor: enable noImplicitAny table.ts

### PR DESCRIPTION
**Pull request by laljikanjareeya**
Fixes #<issue_number_goes_here> (it's a good idea to open an issue first for discussion)

- [ ] Tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: src/table.ts [hunk 1]
```diff
@@ -23,44 +23,35 @@ import * as through from 'through2';
...
-import {SpannerClient as s} from './v1';
 import {PartialResultStream, Row} from './partial-result-stream';
-import {ReadRequest, TimestampBounds} from './transaction';
+import {
+  ReadRequest,
+  TimestampBounds,
+  CommitResponse,
+  ReadResponse,
+  ReadCallback,
+  Transaction,
```
Comment (AVaksman) [inline src/table.ts]: My IDE shows `Transaction` is not used
as well as `ServiceError` on line 21 above

File: src/table.ts [hunk 2]
```diff
@@ -23,44 +23,35 @@ import * as through from 'through2';
...
 export type CreateTableResponse = [
   Table,
   GaxOperation,
   databaseAdmin.longrunning.IOperation
 ];
-type DropTablePromise = Promise<
-  [GaxOperation, databaseAdmin.longrunning.IOperation]
->;
-type ReadPromise = Promise<[Array<Row | Json>]>;
+type DropTableResponse = [GaxOperation, databaseAdmin.longrunning.IOperation];
```
Comment (AVaksman) [inline src/table.ts]: Export it?

File: src/table.ts [hunk 3]
```diff
@@ -23,44 +23,35 @@ import * as through from 'through2';
...
-  (
-    err?: ServiceError | null,
-    table?: Table | null,
-    operation?: GaxOperation | null,
-    apiResponse?: databaseAdmin.longrunning.IOperation | null
-  ): void;
-}
+export type CreateTableCallback = LongRunningCallback<Table>;
 
 type DropTableCallback = ResourceCallback<
```
Comment (AVaksman) [inline src/table.ts]: Export this as well?

File: src/table.ts [hunk 4]
```diff
@@ -23,44 +23,35 @@ import * as through from 'through2';
...
 import {Database} from './database';
-import {SpannerClient as s} from './v1';
 import {PartialResultStream, Row} from './partial-result-stream';
-import {ReadRequest, TimestampBounds} from './transaction';
+import {
+  ReadRequest,
+  TimestampBounds,
+  CommitResponse,
+  ReadResponse,
+  ReadCallback,
```
Comment (AVaksman) [inline src/table.ts:29]: I believe `rows` should be optional =>`rows?` in [`ReadCallback`](https://github.com/laljikanjareeya/nodejs-spanner/blob/4d3d64962a1d4c7fafd2b8610209c7aed6b0585b/src/transaction.ts#L116) interface. You will need to add a few `!`s in system-test as a result of that.

File: src/table.ts [hunk 5]
```diff
@@ -23,44 +23,35 @@ import * as through from 'through2';
 import {Operation as GaxOperation} from 'google-gax';
 import {Json} from './codec';
 import {Database} from './database';
```
Comment (AVaksman) [inline src/table.ts]: You can also import `UpdateSchemaCallback` and `UpdateSchemaResponse` and use it for DropTableCallback/Response below

File: src/table.ts (hunk 2 above)
Comment (AVaksman) [inline src/table.ts]: Suggesting to utilize `UpdateSchemaCallback/Response` as this method is a wrapper around `Database.updateSchema`
```suggestion
export type DropTableResponse = UpdateSchemaResponse;
```

File: src/table.ts (hunk 3 above)
Comment (AVaksman) [inline src/table.ts]: Utilize `UpdateSchemaCallback` as well here
```suggestion
export interface DropTableCallback extends UpdateSchemaCallback{};
```

Optional, I would change [`UpdataSchemaCallback`](https://github.com/laljikanjareeya/nodejs-spanner/blob/4d3d64962a1d4c7fafd2b8610209c7aed6b0585b/src/database.ts#L101-L104)
from
```ts
export type UpdateSchemaCallback = ResourceCallback<
  GaxOperation,
  databaseAdmin.longrunning.IOperation
>;
```
to just 
```ts
export interface UpdateSchemaCallback {
  (
    err: ServiceError | null,
    operation?: GaxOperation | null,
    response?: databaseAdmin.longrunning.IOperation
  ): void;
}
```
as `GaxOperation` is not really a `resource`

File: src/table.ts [hunk 6]
```diff
@@ -298,8 +289,8 @@ class Table {
       callback!
     );
   }
-  deleteRows(keys: Key[]): CommitPromise;
-  deleteRows(keys: Key[], callback: s.CommitCallback): void;
+  deleteRows(keys: Key[]): Promise<CommitResponse>;
```
Comment (AVaksman) [inline src/table.ts]: For consistency you can add
```ts
export interface DeleteRowsCallback extends CommitCallback {}
export type DeleteRowsResponse = CommitResponse;
```
and use them here

File: src/table.ts [hunk 7]
```diff
@@ -391,11 +385,11 @@ class Table {
    *     // Table dropped successfully.
    *   });
    */
-  drop(callback?: DropTableCallback): DropTablePromise | void {
+  drop(callback?: DropTableCallback): Promise<DropTableResponse> | void {
     return this.delete(callback!);
   }
-  insert(rows: object | object[]): CommitPromise;
-  insert(rows: object | object[], callback: s.CommitCallback): void;
+  insert(rows: object | object[]): Promise<CommitResponse>;
```
Comment (AVaksman) [inline src/table.ts]: Same here, for consistency can add and use
```ts
export interface InsertCallback extends CommitCallback {}
export type InsertResponse = CommitResponse;
```

File: src/table.ts [hunk 8]
```diff
@@ -677,12 +671,12 @@ class Table {
    */
   replace(
     rows: object | object[],
-    callback?: s.CommitCallback
-  ): CommitPromise | void {
+    callback?: CommitCallback
```
Comment (AVaksman) [inline src/table.ts]: Same consistency suggestion as above

File: src/table.ts [hunk 9]
```diff
@@ -677,12 +671,12 @@ class Table {
...
+  ): Promise<CommitResponse> | void {
     this._mutate('replace', rows, callback!);
   }
-  update(rows: object | object[]): CommitPromise;
-  update(rows: object | object[], callback: s.CommitCallback): void;
+  update(rows: object | object[]): Promise<CommitResponse>;
```
Comment (AVaksman) [inline src/table.ts]: Same consistency suggestion as above

File: src/table.ts [hunk 10]
```diff
@@ -728,12 +722,12 @@ class Table {
...
     rows: object | object[],
-    callback?: s.CommitCallback
-  ): CommitPromise | void {
+    callback?: CommitCallback
+  ): Promise<CommitResponse> | void {
     this._mutate('update', rows, callback!);
   }
-  upsert(rows: object | object[]): CommitPromise;
-  upsert(rows: object | object[], callback: s.CommitCallback): void;
+  upsert(rows: object | object[]): Promise<CommitResponse>;
```
Comment (AVaksman) [inline src/table.ts]: Same consistency suggestion as above

File: src/table.ts [hunk 11]
```diff
@@ -795,16 +789,17 @@ class Table {
...
     this.database.runTransaction((err, transaction) => {
       if (err) {
         callback(err);
         return;
       }
 
-      transaction[method](this.name, rows);
-      transaction.commit(callback);
+      // tslint:disable-next-line: no-any
+      (transaction as any)![method](this.name, rows);
```
Comment (AVaksman) [inline src/table.ts:809]: can you try without `any`?
```ts
transaction![method](this.name, rows);
```

File: src/table.ts [hunk 12]
```diff
@@ -23,44 +23,35 @@ import * as through from 'through2';
...
+  TimestampBounds,
+  CommitResponse,
+  ReadResponse,
+  ReadCallback,
+  Transaction,
+  CommitCallback,
+} from './transaction';
 import {google as databaseAdmin} from '../proto/spanner_database_admin';
-import {Schema, ResourceCallback} from './common';
+import {Schema, ResourceCallback, LongRunningCallback} from './common';
```
Comment (AVaksman) [inline src/table.ts]: Seems if you utilize `UpdataSchemaCallback/Response`, you do not need to import `ResourceCallback` here

**Review by AVaksman — CHANGES_REQUESTED** [review]

File: src/table.ts (hunk 4 above)
Comment (laljikanjareeya) [reply src/table.ts:29]: Used `NormalCallback<T>`

File: src/table.ts (hunk 3 above)
Comment (laljikanjareeya) [reply src/table.ts]: Dave Suggested in `database.ts` PR to use `ResourceCallback` for same.

File: src/table.ts (hunk 11 above)
Comment (laljikanjareeya) [reply src/table.ts:809]: while enable `NoImplicitAny` it gives an error.

File: src/table.ts (hunk 3 above)
Comment (AVaksman) [reply src/table.ts]: ok

File: src/table.ts (hunk 11 above)
Comment (AVaksman) [reply src/table.ts:809]: ok

**Review by AVaksman — APPROVED** [review]

**Pull request closed without merging**